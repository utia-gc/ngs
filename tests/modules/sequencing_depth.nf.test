nextflow_process {

    name "Test Process sequencing_depth"
    script "modules/sequencing_depth.nf"
    process "sequencing_depth"

    test("sequencing_depth writes file of sequencing depth -- SE reads; one read") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            process {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        readType:     "${readsLane1.getReadType()}",
                        trimStatus:   "raw"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                    ],
                    [
                        file("${readsLane1.getR2()}"),
                    ]
                ]
                input[1] = file("${refs.getGenomeIndex()}")
                """
            }
        }

        then {
            // test process success
            assert process.success

            // test expected sequencing depth file made
            with(process.out.depth.get(0)) {
                assert size() == 2
                // test metadata
                with(get(0)) {
                    assert sampleName == "${readsLane1.getSampleName()}"
                    assert readType   == "${readsLane1.getReadType()}"
                    assert trimStatus == "raw"
                }
                // test sequencing depth reported
                assert get(1).isNumber()
                assert get(1).toDouble() >= 0
            }

            // test reproducibility
            assert snapshot(process.out).match()
        }

    }

    test("sequencing_depth writes file of sequencing depth -- SE reads; multiple reads") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()
        def readsLane2 = new ReadsSELane2()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            process {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        readType:     "${readsLane1.getReadType()}",
                        trimStatus:   "raw"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                        file("${readsLane2.getR1()}"),
                    ],
                    [
                        file("${readsLane1.getR2()}"),
                        file("${readsLane2.getR2()}"),
                    ]
                ]
                input[1] = file("${refs.getGenomeIndex()}")
                """
            }
        }

        then {
            // test process success
            assert process.success

            // test expected sequencing depth file made
            with(process.out.depth.get(0)) {
                assert size() == 2
                // test metadata
                with(get(0)) {
                    assert sampleName == "${readsLane1.getSampleName()}"
                    assert readType   == "${readsLane1.getReadType()}"
                    assert trimStatus == "raw"
                }
                // test sequencing depth reported
                assert get(1).isNumber()
                assert get(1).toDouble() >= 0
            }

            // test reproducibility
            assert snapshot(process.out).match()
        }

    }

    test("sequencing_depth writes file of sequencing depth -- PE reads; one read-pair") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            process {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        readType:     "${readsLane1.getReadType()}",
                        trimStatus:   "raw"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                    ],
                    [
                        file("${readsLane1.getR2()}"),
                    ]
                ]
                input[1] = file("${refs.getGenomeIndex()}")
                """
            }
        }

        then {
            // test process success
            assert process.success

            // test expected sequencing depth file made
            with(process.out.depth.get(0)) {
                assert size() == 2
                // test metadata
                with(get(0)) {
                    assert sampleName == "${readsLane1.getSampleName()}"
                    assert readType   == "${readsLane1.getReadType()}"
                    assert trimStatus == "raw"
                }
                // test sequencing depth reported
                assert get(1).isNumber()
                assert get(1).toDouble() >= 0
            }

            // test reproducibility
            assert snapshot(process.out).match()

        }

    }

    test("sequencing_depth writes file of sequencing depth -- PE reads; multiple read-pairs") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()
        def readsLane2 = new ReadsPELane2()
        def refs       = new RefsEnsembl()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            process {
                """
                input[0] = [
                    [
                        sampleName:   "${readsLane1.getSampleName()}",
                        readType:     "${readsLane1.getReadType()}",
                        trimStatus:   "raw"
                    ],
                    [
                        file("${readsLane1.getR1()}"),
                        file("${readsLane2.getR1()}"),
                    ],
                    [
                        file("${readsLane1.getR2()}"),
                        file("${readsLane2.getR2()}"),
                    ]
                ]
                input[1] = file("${refs.getGenomeIndex()}")
                """
            }
        }

        then {
            // test process success
            assert process.success

            // test expected sequencing depth file made
            with(process.out.depth.get(0)) {
                assert size() == 2
                // test metadata
                with(get(0)) {
                    assert sampleName == "${readsLane1.getSampleName()}"
                    assert readType   == "${readsLane1.getReadType()}"
                    assert trimStatus == "raw"
                }
                // test sequencing depth reported
                assert get(1).isNumber()
                assert get(1).toDouble() >= 0
            }

            // test reproducibility
            assert snapshot(process.out).match()

        }

    }

}
