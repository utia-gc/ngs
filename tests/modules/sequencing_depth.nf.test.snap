{
    "sequencing_depth writes file of sequencing depth -- SE reads": {
        "content": [
            {
                "0": [
                    "SRR1066657_S3_L001_seq-depth.txt:md5,afcf5b26783b1e7189a66c83073e0aa4"
                ],
                "depth": [
                    "SRR1066657_S3_L001_seq-depth.txt:md5,afcf5b26783b1e7189a66c83073e0aa4"
                ]
            }
        ],
        "timestamp": "2023-09-28T19:33:40+0000"
    },
    "sequencing_depth writes file of sequencing depth -- SE reads no lane info": {
        "content": [
            {
                "0": [
                    "SRR1066657_seq-depth.txt:md5,afcf5b26783b1e7189a66c83073e0aa4"
                ],
                "depth": [
                    "SRR1066657_seq-depth.txt:md5,afcf5b26783b1e7189a66c83073e0aa4"
                ]
            }
        ],
        "timestamp": "2023-09-28T19:33:40+0000"
    },
    "sequencing_depth writes file of sequencing depth -- PE reads": {
        "content": [
            {
                "0": [
                    "SRR6924569_S1_L001_seq-depth.txt:md5,c6b428ff5b8f681e71ec3b138849fec1"
                ],
                "depth": [
                    "SRR6924569_S1_L001_seq-depth.txt:md5,c6b428ff5b8f681e71ec3b138849fec1"
                ]
            }
        ],
        "timestamp": "2023-09-28T19:33:40+0000"
    }
}